File: Learn/30-Days-Of-Code/Day 24/morelinked.cpp
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node *next;
        Node (int d) {
            data = d;
            next = NULL;
        }
};

class Solution {
    public:
        Node * removeDuplicates(Node *head) {
            Node **curr = &head;
            while (*curr && (*curr)->next) {
                if ((*curr)->data == (*curr)->next->data) {
                    Node *t = (*curr)->next;
                    (*curr)->next = (*curr)->next->next;
                    free(t);
                    continue;
                }
                curr = &(*curr)->next;
            }
            return head;
        }
        Node * insert(Node *head, int data) {
            Node *p = new Node(data);
            Node **curr = &head;
            while (*curr) {
                curr = &(*curr)->next;
            }
            *curr = p;
            return head;
        }
        void display(Node *head) {
            Node *curr = head;
            while (curr) {
                cout << curr->data << " ";
                curr = curr->next;
            }
        }
};

int main() {
    Node *head = NULL;
    Solution mylist;
    int T, data;
    cin >> T;
    while (T-- > 0) {
        cin >> data;
        head = mylist.insert(head, data);
    }
    head = mylist.removeDuplicates(head);
    mylist.display(head);
    return 0;
}

File: Learn/30-Days-Of-Code/Day 24/morelinked.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def insert(self, head, data):
        p = Node(data)
        if head is None:
            head = p
        elif head.next is None:
            head.next = p
        else:
            start = head
            while start.next is not None:
                start = start.next
            start.next = p
        return head
    def display(self, head):
        curr = head
        while curr:
            print(curr.data, end=' ')
            curr = curr.next
    def removeDuplicates(self, head):
        curr = head
        while curr and curr.next:
            if curr.data == curr.next.data:
                curr.next = curr.next.next
                continue
            curr = curr.next
        return head

mylist = Solution()
T = int(input())
head = None
for i in range(T):
    data = int(input())
    head = mylist.insert(head, data)
head = mylist.removeDuplicates(head)
mylist.display(head)
